<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Quick start · Escape.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://HAMdetector.github.io/MyPkg2.jl/quick_start/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>Escape.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Quick start</a><ul class="internal"><li><a class="tocitem" href="#Running-HAMdetector"><span>Running HAMdetector</span></a></li><li><a class="tocitem" href="#Providing-an-HLA-annotation-file"><span>Providing an HLA annotation file</span></a></li><li><a class="tocitem" href="#Convergence-diagnostics"><span>Convergence diagnostics</span></a></li><li><a class="tocitem" href="#Posterior-predictive-checks"><span>Posterior predictive checks</span></a></li><li><a class="tocitem" href="#Parsing-HLA-alleles"><span>Parsing HLA alleles</span></a></li></ul></li><li><a class="tocitem" href="../advanced_usage/">Advanced usage</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Quick start</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Quick start</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="github.com/HAMdetector/Escape.jl.git" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Quick-start"><a class="docs-heading-anchor" href="#Quick-start">Quick start</a><a id="Quick-start-1"></a><a class="docs-heading-anchor-permalink" href="#Quick-start" title="Permalink"></a></h1><h2 id="Running-HAMdetector"><a class="docs-heading-anchor" href="#Running-HAMdetector">Running HAMdetector</a><a id="Running-HAMdetector-1"></a><a class="docs-heading-anchor-permalink" href="#Running-HAMdetector" title="Permalink"></a></h2><p>HAMdetector requires two things: A set of aligned amino acid sequences that are annotated  with HLA alleles and a phylogenetic tree. The sequence alignment has to be provided in FASTA format, the phylogenetic tree in Newick format.</p><p>The easiest way of running HAMdetector is by providing the input files in a standardized way, that is, the sequences in the .fasta file are annotated by an underscore separated list of HLA alleles and the tips of the phylogenetic tree are labeled by their position in the multiple sequence alignment.</p><p>An example of valid input files is shown below:</p><p><strong>alignment.fasta</strong></p><pre><code class="language-txt">&gt;HLA-A*32_HLA-A*03_HLA-B*52_HLA-B*14_HLA-C*16_HLA-C*16
MG-RASVMG-RA*V
&gt;HLA-A*68_HLA-A*31_HLA-B*44_HLA-B*08_HLA-C*07_HLA-C*07
MGARASVMG-RT*V
&gt;HLA-A*24_HLA-A*24_HLA-B*15_HLA-B*46_HLA-C*12_HLA-C*12
MSARASVMGSRSSV
&gt;HLA-A*02_HLA-A*03_HLA-B*58_HLA-B*44_HLA-C*14_HLA-C*06
MSARASVMG-RRSV
&gt;HLA-A*03_HLA-A*31_HLA-B*55_HLA-B*40_HLA-C*16_HLA-C*03
MSARASVMG-RMSV</code></pre><p><strong>phylogeny.tree</strong></p><pre><code class="language-txt">((2:0.033310,1:0.118635):33.615353,(3:0.107192,5:0.857526):0.102147,4:0.247706);</code></pre><p>Assuming these files were saved as <code>/home/user/Desktop/alignment.fasta</code> and <code>/home/user/Desktop/phylogeny.tree</code>, HAMdetector can be run with the following commands:</p><pre><code class="language-julia">using Escape

data = HLAData(
    alignment_file = &quot;/home/user/Desktop/alignment.fasta&quot;,
    tree_file = &quot;/home/user/Desktop/phylogeny.tree&quot;
)

result = run_model(data)

summary = replacement_summary(result)</code></pre><p>In a first step, the input data are converted into a suitable format to run the HAMdetector model. This step includes calculating phylogeny information and running MHCFlurry for epitope prediction.</p><p>In a second step, the HAMdetector model is run on the input data.</p><p>The <code>replacement_summary</code> function then computes a table of all observed replacements, sorted by decreasing estimated probability of being HLA allele associated. The column <em>posterior_p</em> denotes the posterior probability of the HLA regression coefficient being larger than 0, conditioned on the model and the observed data. A posterior probability of 1 denotes that the model is certain that a given replacement is an HLA-associated mutation, whereas a posterior probability of 0.5 means that there is no evidence of the replacement being HLA-associated (0.5 because a regression coefficient centered around 0 allocates equal probability to negative and positive values). </p><p>The returned structure is a <code>DataFrame</code> object and elements can be accessed using standard indexing, e.g. <code>summary[1:10, :posterior_p]</code> returns the posterior probabilities of the first ten entries.</p><p>DataFrames can be saved as .csv files using the CSV package:</p><pre><code class="language-julia">using CSV

CSV.write(&quot;/home/user/Desktop/summary.csv&quot;, summary)</code></pre><p>The <code>HLAData</code> and <code>run_model</code> functions have an optional keyword argument <code>save_file</code> that can be used to save the converted input data and the result object.</p><pre><code class="language-julia">data = HLAData(
    alignment_file = &quot;/home/user/Desktop/alignment.fasta&quot;,
    tree_file = &quot;/home/user/Desktop/phylogeny.tree&quot;,
    save_file = &quot;/home/user/Desktop/input_data.jld2&quot;
)

result = run_model(data, save_file = &quot;/home/user/Desktop/model_result.jld2&quot;)

summary = replacement_summary(result)</code></pre><p>The file extension &quot;.jld2&quot; is used because HAMdetector uses the <a href="https://github.com/JuliaIO/JLD2.jl">JLD2</a> package for saving and loading data.</p><p>To restore saved results from disk, use the <code>load_data</code> and <code>load_result</code> functions:</p><pre><code class="language-julia">using Escape

result = load_result(&quot;/home/user/Desktop/model_result.jld2&quot;)
data = load_data(&quot;/home/user/Desktop/input_data.jld2&quot;)</code></pre><h2 id="Providing-an-HLA-annotation-file"><a class="docs-heading-anchor" href="#Providing-an-HLA-annotation-file">Providing an HLA annotation file</a><a id="Providing-an-HLA-annotation-file-1"></a><a class="docs-heading-anchor-permalink" href="#Providing-an-HLA-annotation-file" title="Permalink"></a></h2><p>For some datasets, sequences are not annotated with HLA information directly, but with a unique sequence identifier. For these cases, the <code>HLAData</code> function accepts an additional argument to provide a file that matches these sequence identifiers with the HLA annotation. The HLA annotation file has to be a comma-, tab- or semicolon-separated list of values with the columns &quot;identifier&quot;, &quot;A1&quot;, &quot;A2&quot;, &quot;B1&quot;, &quot;B2&quot;, &quot;C1&quot;, &quot;C2&quot;. An example of a valid annotation file is shown below:</p><p><strong>hla_annotations.csv</strong></p><pre><code class="language-text">idenfier,A1,A2,B1,B2,C1,C2
seq1,HLA-A*74,HLA-A*24,HLA-B*44,HLA-B*40,HLA-C*04,HLA-C*02
seq2,HLA-A*11,HLA-A*11,HLA-B*51,HLA-B*07,HLA-C*01,HLA-C*01
seq3,HLA-A*11,HLA-A*11,HLA-B*15,HLA-B*35,HLA-C*05,HLA-C*07
seq4,HLA-A*24,HLA-A*23,HLA-B*54,HLA-B*41,HLA-C*06,HLA-C*05
seq5,HLA-A*31,HLA-A*01,HLA-B*57,HLA-B*35,HLA-C*01,HLA-C*06</code></pre><p>Note that, as with providing HLA annotation directly in the .fasta file, the order of the individual HLA alleles does not matter, e.g. a sequence annotated with HLA-B*18 in the B1 column and HLA-B*51 in the B2 column is treated the same as an otherwise identical sequence annotated with HLA-B*51 in the B1 column and HLA-B*18 in the B2 column.</p><p>HLA alleles do not have to be provided according to the standard HLA nomenclature, as HAMdetector is able to parse a broad range of commonly used HLA formats, also see section <a href="##parsing-hla-alleles">Parsing HLA section</a>.</p><pre><code class="language-julia">using Escape

data = HLAData(
    alignment_file = &quot;/home/user/Desktop/alignment_with_ids.fasta&quot;,
    tree_file = &quot;/home/user/Desktop/phylogeny.tree&quot;,
    hla_annotation_file = &quot;/home/user/Desktop/hla_annotations.csv&quot;
)</code></pre><p>If an HLA annotation file is provided, the tips of the phylogenetic tree may either be labeled by the position of the corresponding sequence in the multiple sequence alignment - as explained previously- or by the unique sequence identifier.</p><h2 id="Convergence-diagnostics"><a class="docs-heading-anchor" href="#Convergence-diagnostics">Convergence diagnostics</a><a id="Convergence-diagnostics-1"></a><a class="docs-heading-anchor-permalink" href="#Convergence-diagnostics" title="Permalink"></a></h2><p>HAMdetector uses CmdStan for model fitting, which in turn uses a variant of Hamiltonian Monte Carlo sampling to generate samples from the typical set.  To check that these samples faithfully reflect the posterior distribution, HAMdetector uses CmdStan&#39;s diagnostic tool. The output can be retrieved with:</p><pre><code class="language-julia">using Escape

Escape.diagnostics(result)</code></pre><p>The output is not nicely formated yet, but a successful model run should end in &quot;Processing complete, no problems detected&quot;.</p><h2 id="Posterior-predictive-checks"><a class="docs-heading-anchor" href="#Posterior-predictive-checks">Posterior predictive checks</a><a id="Posterior-predictive-checks-1"></a><a class="docs-heading-anchor-permalink" href="#Posterior-predictive-checks" title="Permalink"></a></h2><p>Bayesian statistics allows for a form of model checking called <a href="https://arxiv.org/abs/1709.01449">posterior predictive checks</a>. A simple form of posterior predictive checks are calibration plots, which can be generated with:</p><pre><code class="language-julia">using Escape, Plots

Escape.calibration_plot(result)</code></pre><p>For a well calibrated model, the vertical bars should span the diagonal line. Strong deviations from this suggest that something went wrong. A similar plot that specifically checks calibration of the internally calculated phylogeny can be generated with <code>Escape.phylogeny_calibration(result)</code>.</p><p>Based on the quality of the provided phylogenetic tree, the calibration plot might show stronger deviations. This is usually not an issue because HAMdetector is able to handle noisy phylogenetic inputs.</p><h2 id="Parsing-HLA-alleles"><a class="docs-heading-anchor" href="#Parsing-HLA-alleles">Parsing HLA alleles</a><a id="Parsing-HLA-alleles-1"></a><a class="docs-heading-anchor-permalink" href="#Parsing-HLA-alleles" title="Permalink"></a></h2><p>For many datasets, HLA annotations do not follow the <a href="http://hla.alleles.org/nomenclature/naming.html">official HLA nomenclature</a>. HAMdetector allows parsing of a wide range of HLA allele notations and tries to be as flexible as possible. To check if HAMdetector is able to parse HLA allele annotations in your dataset, use the <code>parse_allele</code> function:</p><pre><code class="language-julia">using Escape

Escape.parse_allele(&quot;B51&quot;)
Escape.parse_allele(&quot;B5101&quot;)
Escape.parse_allele(&quot;A*0103&quot;)
Escape.parse_allele(&quot;Cw*0302&quot;)
Escape.parse_allele(&quot;HLA-B51&quot;)</code></pre><p>Please note that HAMdetector right now supports either 2-digit or 4-digit HLA alleles, but no mixture of both. This limitation can in principle be solved by applying a technique called <a href="http://www.stat.columbia.edu/~gelman/research/published/multi2.pdf">partial pooling</a>,  which is going to be implemented in future release of HAMdetector.</p><p>Missing HLA alleles are supported by providing a blank HLA entry. Internally, missing HLA alleles are not imputed, but treated as 0 entries in the HLA matrix. This means that any effect that cannot be explained by one of the present HLA alleles is accounted for by the error term (the intercept), which is a reasonable default if HLA alleles are missing at random.</p></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="../advanced_usage/">Advanced usage »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.3 on <span class="colophon-date" title="Sunday 1 October 2023 00:22">Sunday 1 October 2023</span>. Using Julia version 1.6.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
